Zcoin atomic swap to be merged

devwarrior777 opened the pull request:
### An atomic swap has been done between Decred testnet & Zcoin testnet

The output of the swap has been pasted in the file below:

There remain some issues with `estimatefee` and/or `estimatesmartfee` but these can probably be resolved either with talked about changes to the atomic swap **cmd** code or in the xzcwallet/xzcd zcoin-specific code. The swap is nonetheless successful.

To view the transactions on the Zcoin blockchain please use the temporary explorer at:
http://45.76.2.255

API Example: http://45.76.2.255/api/getrawtransaction?txid=3583e1b9f2188862f0648e54ea29c4c2d8e8f31dfb015be80f18918cd106eb3f&decrypt=1

Note: `warning: falling back to mempool relay fee policy`

[xcat-xzc-dcr_2018_01_13.txt](https://github.com/decred/atomicswap/files/1631881/xcat-xzc-dcr_2018_01_13.txt)

devwarrior777 commented:
xcat-xzc-dcr_2018_01_13.txt file attached above shows the test output.

devwarrior777 commented:
For the `estimatefee` enhaancement I have opened and issue with zcoin ...

devwarrior777 commented:
I have opened an Issue to get support for `estimatefee` by Zcoin.
https://github.com/zcoinofficial/zcoin/issues/170
This RPC will almost certainly be in the next Zcoin release

jrick commented:
I didn't go digging through their source, but it's possible (especially on the lesser used testnet) that these are implemented, but always return -1 due to a lack of transactions.  When I was testing support for partatomicswap we had to flood the network with other transactions in order for the estimate RPCs to return anything useful.

devwarrior777 commented:
I have done a similar successful swap on mainnet and attach the details below.
- On mainnet I had to change the Transaction version to v1:  01000000 due to checks for CSV readiness made on mainnet in bitcoin 0.13 that are not made in testnet. So there will be at least one more commit for that.
- Even on mainnet the `estimatefee` API still returns -1 .. a fulltime Zcoin dev is also looking at this issue and has promised changes for the next release.
- The `relayfee` value returned from `getinfo` is 0.001 so this is what we pass as a parameter to `fundrawtransaction` so I will investigate from there ;) It looks like similar behavior on mainnet.

Thanks for your feedback on both `estimatefee` and this Contract fee anomaly

**mainnet test**  [xzc_dcr_mainnet.txt](https://github.com/decred/atomicswap/files/1647413/xzc_dcr_mainnet.txt)

__mainnet explorer:__ https://chainz.cryptoid.info/xzc/

devwarrior777 commented:
I owe you guys a refund proof:

[xzc_dcr_mainnet_refund_2018_01_24__1330.txt](https://github.com/decred/atomicswap/files/1669901/xzc_dcr_mainnet_refund_2018_01_24__1330.txt)

devwarrior777 commented:
Added 1 more commit:
Changes for "secret size attack changes"

devwarrior777 commented:
Added 1 commit:
"[main] update for secret size attack on cross-chain hash time lock sm…"
Updated deps

devwarrior777 commented:
Decred atomicswap
--------------------------
- Added 1 commit:
"[main] update for secret size attack on cross-chain hash time lock sm… "
- Updated [xczd] for AtomicSwapDataPushes
https://github.com/zcoinofficial/xzcd/blob/ae307d1ae9070c5434a6a203992db0ad35712e96/txscript/standard.go#L630
- Updated all the `vendor` dependencies

Zcoin
--------
Zcoin v0.13.5.7  "`fundrawtransaction` now applies fees"

`estimatefee/estimatesmartfee` - all code exists and is enabled but the # transactions / block is not considered statistically significant. Experimenting with changing bitcoin parameters shows that estimation of fees works and will be available when transaction volumes increase.

jrick commented:
Can you rebase+squash this and resolve the conflicts with the Gopkg.toml/lock files?  To resolve those conflicts, you can revert both back to their state on master, add your new deps to Gopkg.toml, and then re-run dep ensure to update the lock without modifying anything else already in the lock.

devwarrior777 commented:
This is getting in a state - close this and I will do a new PR
